### PR TITLE
Support path mappings in JacocoLCOVFormatter

### PR DESCRIPTION
**Pull request by dorranh**
Addresses #12159 by adding support for an additional path format in the JacocoCoverageRunner's `-paths-for-coverage.txt` file which allows rule authors to provide either individual source paths or a new source to class path mapping: `srcPath///classPath`. This new format provides a means to handle cases where a source file's package hierarchy is different than the directory tree where it resides, which is fairly common in other JVM languages like Scala (see #12159 for examples). 

Since `JacocoLCOVFormatter` still supports the old path format, this change should not affect any downstream users of `JacocoCoverageRunner`. The goal is just to provide a mechanism for libraries like `rules_scala` to be able to implement coverage support for source files with varying directory structures. 

cc @comius @sjoerdvisscher @liucijus

### CODE REVIEW

File: src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java [hunk 1]
```diff
@@ -58,6 +30,19 @@
 import org.jacoco.report.ISourceFileLocator;
 import sun.misc.Unsafe;
 
+import java.io.*;
```
Comment (c-mita) [inline src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java:33]: This import cleanup falls afoul of our style guide (no wildcard imports).

File: src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    ImmutableSet<String> execPaths = ImmutableSet.of(
+      "/parent/dir/com/example/Foo.java"
+    );
+    JacocoLCOVFormatter formatter = new JacocoLCOVFormatter(execPaths);
+    IReportVisitor visitor = formatter.createVisitor(new PrintWriter(writer), new TreeMap<String, BranchCoverageDetail>());
+    try {
+      visitor.visitBundle(mockBundle, mock(ISourceFileLocator.class));
+      visitor.visitEnd();
+    } catch (IOException e) {
+      Assert.fail(Throwables.getStackTraceAsString(e));
```
Comment (c-mita) [inline src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java:93]: Why catch the IOException to then just "Assert.fail"? Just let the IOException bubble up and terminate the test.

File: src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.TreeMap;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.Mockito.*;
```
Comment (c-mita) [inline src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java:39]: Wildcard imports are against the Google style guide.

File: src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java (hunk 2 above)
Comment (dorranh) [reply src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java]: Makes sense, I've updated the tests accordingly.

**Comment by dorranh** [comment]
Hi @c-mita, thanks for your review - I've made the suggested changes. Please let me know if there is anything else you need on my end.

**Review by c-mita — APPROVED** [review]
LGTM - sending for internal review.

File: src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java [hunk 4]
```diff
@@ -130,24 +133,26 @@ public void create() throws IOException {
 
   @VisibleForTesting
   void createReport(
-      final IBundleCoverage bundleCoverage, final Map<String, BranchCoverageDetail> branchDetails)
-      throws IOException {
+    final IBundleCoverage bundleCoverage, final Map<String, BranchCoverageDetail> branchDetails)
+    throws IOException {
     JacocoLCOVFormatter formatter = new JacocoLCOVFormatter(createPathsSet());
-    final IReportVisitor visitor = formatter.createVisitor(reportFile, branchDetails);
+    try (PrintWriter writer = new PrintWriter(java.nio.file.Files.newBufferedWriter(reportFile.toPath(), UTF_8, CREATE, APPEND))
```
Comment (c-mita) [inline src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java:139]: Style nit: Use an import to get `newBufferedWriter`.

File: src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java [hunk 5]
```diff
@@ -252,7 +258,11 @@ private void analyzeUninstrumentedClassesFromJar(
    * Adds to the given {@link Set} the paths found in a txt file inside the given jar.
    *
    * <p>If a jar contains uninstrumented classes it will also contain a txt file with the paths of
-   * each of these classes, one on each line.
+   * each of these classes, called "-paths-for-coverage.txt". This file expects one path per line,
+   * specified as either:
+   * 1. A single path (e.g. /dir/com/example/Foo.java)
+   * 2. A mapping between source and class paths delimited with by ///
+   * (e.g. /dir/Foo.java////com/example/Foo.java)
```
Comment (c-mita) [inline src/java_tools/junitrunner/java/com/google/testing/coverage/JacocoCoverageRunner.java:265]: Style: Reformat using `<ul>` tags.

File: src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      "/parent/dir/com/example/Foo.java"
+    );
+    JacocoLCOVFormatter formatter = new JacocoLCOVFormatter(execPaths);
+    IReportVisitor visitor = formatter.createVisitor(new PrintWriter(writer), new TreeMap<String, BranchCoverageDetail>());
+    visitor.visitBundle(mockBundle, mock(ISourceFileLocator.class));
+    visitor.visitEnd();
+    String coverageOutput = writer.toString();
+    for (String sourcePath : execPaths) {
+      assertThat(coverageOutput).contains(sourcePath);
+    }
```
Comment (c-mita) [inline src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java:95]: Style: Separate out the test cases with empty lines into "arrange, act, assert" blocks.

File: src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  @Test
+  public void testVisitBundleWithMappedUnixPath() throws IOException {
+    // Paths
+    String srcPath = "/some/other/dir/Foo.java";
+    ImmutableSet<String> execPaths = ImmutableSet.of(
+      String.format("%s%s%s", srcPath, JacocoLCOVFormatter.EXEC_PATH_DELIMITER, "/com/example/Foo.java")
```
Comment (c-mita) [inline src/java_tools/junitrunner/javatests/com/google/testing/coverage/JacocoLCOVFormatterUninstrumentedTest.java:119]: Style: `EXEC_PATH_DELIMITER` can be inlined here. The pattern is essentially API and it makes this test clearer to have a simple string.
This also means `JacocoLCOVFormatter.EXEC_PATH_DELIMTER` can be made `private`, not `public`.

**Review by c-mita — COMMENTED** [review]
A few minor stylistic points came up in internal review. Since they weren't anything functional, I expedited the process by applying the changes myself so you don't need to do anything.

I've left a few review comments reflecting our internal readability review for reference though, as they may be helpful for future PRs.

**Comment by dorranh** [comment]
Hi @c-mita, thanks for helping with the additional comments! I will keep the review points in mind for future PRs.

**Pull request closed without merging**